README.md
# StoReS

Storage of Reionization Simulations

## Package details

The package provides methods for retrieving simulations of reionization and cosmic dawn. Full documentation can be found [here](https://stores.readthedocs.io/en/latest/).

## INSTALLATION

One can install a stable version of this package using pip by running the following command::

    pip install StoReS

In order to use the latest version, one can clone this package running the following::

    git clone https://github.com/sambit-giri/StoReS.git

To install the package in the standard location, run the following in the root directory::

    python setup.py install

In order to install it in a separate directory::

    python setup.py install --home=directory

One can also install it using pip by running the following command::

    pip install git+https://github.com/sambit-giri/StoReS.git

The dependencies should be installed automatically during the installation process. If they fail for some reason, you can install them manually before installing StoReS. The list of required packages can be found in the requirements.txt file present in the root directory.

### Tests

For testing, one can use [pytest](https://docs.pytest.org/en/stable/) or [nosetests](https://nose.readthedocs.io/en/latest/). Both packages can be installed using pip. To run all the test script, run the either of the following::

    python -m pytest tests
    
	nosetests -v

## USAGE

Script to get C2Ray reionization simulations.

```python
import numpy as np 
import matplotlib.pyplot as plt 
import pickle
from glob import glob
from StoReS import * 
import tools21cm as t2c 

box_len  = 244/0.7
save_dir = '../../'

xf_url = 'https://ttt.astro.su.se/~gmell/244Mpc/LB1/'
dn_url = 'https://ttt.astro.su.se/~gmell/244Mpc/densities/nc250/coarser_densities/'
url_dict = {'dens': dn_url, 'xfrac': xf_url}

c2r = C2RAY(work_dir=save_dir)
c2r.set_links(url_dict)
z = 6.905
data = c2r.get_data_z(z)
xf = data['xfrac']
dn = data['dens']


fig, axs = plt.subplots(1,2,figsize=(13,6))
dt = t2c.calc_dt(xf, dn, z)
ps, ks = t2c.power_spectrum_1d(dt, kbins=20, box_dims=box_len)
axs[0].pcolormesh(np.linspace(0,box_len,dt.shape[0]+1), 
                  np.linspace(0,box_len,dt.shape[1]+1), 
                  dt[:,:,0],
                  cmap='jet')
axs[0].set_xlabel('[Mpc]', fontsize=13)
axs[0].set_ylabel('[Mpc]', fontsize=13)
axs[1].loglog(ks, ps*ks**3/2/np.pi**2, lw=3, 
    label='$z={:.3f}$, $x_\mathrm{{HII}}={:.1f}$'.format(z,xf.mean()))
axs[1].legend(fontsize=14)
axs[1].set_xlabel('$k$ [1/Mpc]', fontsize=13)
axs[1].set_ylabel('$k^3 P/(2\pi^2)$ [mK$^2$]', fontsize=13)
plt.tight_layout()
plt.show()

```

<img src="images/LB1_z6p905.png" width="600">

Script to get (re)ionization history of the LB1 simulation from Ref. [[1]](#1).

```python
import numpy as np 
import matplotlib.pyplot as plt 
import pickle, os
from glob import glob
from StoReS import * 
import tools21cm as t2c 
from tqdm import tqdm 

box_len  = 244/0.7
save_dir = '../../'
eor_hist = {}

xf_url = 'https://ttt.astro.su.se/~gmell/244Mpc/LB1/'
dn_url = 'https://ttt.astro.su.se/~gmell/244Mpc/densities/nc250/coarser_densities/'
url_dict = {'xfrac': xf_url, 'dens': dn_url, }
c2r = C2RAY(work_dir=save_dir)
c2r.set_links(url_dict)
zs = np.intersect1d(c2r.zs_dict['xfrac'], c2r.zs_dict['dens'])
xs = []
for i,z in tqdm(enumerate(zs)):
    # z = 6.905
    data = c2r.get_data_z(z)
    xf = data['xfrac']
    dn = data['dens']
    xs.append(xf.mean())
    os.remove(data['xfrac_filename'])
    os.remove(data['dens_filename'])
xs = np.array(xs)
eor_hist['LB1'] = {'z': zs, 'x': xs}

fig, ax = plt.subplots(1,1,figsize=(6,5))
fig.suptitle('Dixon et al. (2016)')
ax.plot(eor_hist['LB1']['z'], eor_hist['LB1']['x'], lw=3, label='LB1')
ax.set_xlabel('z', fontsize=14)
ax.set_ylabel('$x_\mathrm{HII}$', fontsize=14)
ax.legend()
ax.axis([6,14,-0.01,1.01])
plt.tight_layout()
plt.show()
```
<img src="images/EoR_history_LB1.png" width="600">


## CONTRIBUTING

If you find any bugs or unexpected behavior in the code, please feel free to open a [Github issue](https://github.com/sambit-giri/StoReS/issues). The issue page is also good if you seek help or have suggestions for us. 

## References
<a id="1">[1]</a> 
Dixon, K. L., Iliev, I. T., Mellema, G., Ahn, K., & Shapiro, P. R. (2016). The large-scale observational signatures of low-mass galaxies during reionization. Monthly Notices of the Royal Astronomical Society, 456(3), 3011-3029. [arXiv:1512.03836](https://arxiv.org/abs/1512.03836).


docs/changelog.rst
=========
Changelog
=========

v0.1
----
* Initial version of the package. 
* It was created to retrieve and manipulate |c2ray|_ and |cubep3m|_ simulations.

.. |c2ray| replace:: C\ :sup:`2`\RAY
.. _c2ray: https://github.com/garrelt/C2-Ray3Dm

.. |cubep3m| replace:: CUBEP\ :sup:`3`\M
.. _cubep3m: https://wiki.cita.utoronto.ca/index.php/CubePM


examples/test.py
import numpy as np 
import matplotlib.pyplot as plt 
import pickle
from glob import glob
from StoReS import * 
import tools21cm as t2c 

box_len  = 244/0.7
save_dir = '../../'

xf_url = 'https://ttt.astro.su.se/~gmell/244Mpc/LB1/'
dn_url = 'https://ttt.astro.su.se/~gmell/244Mpc/densities/nc250/coarser_densities/'
url_dict = {'dens': dn_url, 'xfrac': xf_url}

c2r = C2RAY(work_dir=save_dir)
c2r.set_links(url_dict)
z = 6.905
data = c2r.get_data_z(z)
xf = data['xfrac']
dn = data['dens']

# dn_file_link = 'https://ttt.astro.su.se/~gmell/244Mpc/densities/nc250/coarser_densities/6.905n_all.dat'
# xf_file_link = 'https://ttt.astro.su.se/~gmell/244Mpc/LB1/xfrac3d_6.905.bin'
# dn_filename = download_simulation(dn_file_link, save_dir, verbose=True, overwrite=False)
# xf_filename = download_simulation(xf_file_link, save_dir, verbose=True, overwrite=False)
# print(dn_filename, xf_filename)
# dn = t2c.DensityFile(dn_filename).cgs_density
# xf = t2c.XfracFile(xf_filename).xi

print(dn.shape, xf.shape)

# fig, axs = plt.subplots(1,2,figsize=(13,6))
# axs[0].pcolormesh(np.linspace(0,box_len,dn.shape[0]+1), 
# 				  np.linspace(0,box_len,dn.shape[1]+1), 
# 				  dn[:,:,0],
# 				  cmap='jet')
# axs[0].set_xlabel('[Mpc]', fontsize=13)
# axs[0].set_ylabel('[Mpc]', fontsize=13)
# axs[1].pcolormesh(np.linspace(0,box_len,xf.shape[0]+1), 
# 				  np.linspace(0,box_len,xf.shape[1]+1), 
# 				  xf[:,:,0],
# 				  cmap='jet')
# axs[1].set_xlabel('[Mpc]', fontsize=13)
# axs[1].set_ylabel('[Mpc]', fontsize=13)
# plt.tight_layout()
# plt.show()


fig, axs = plt.subplots(1,2,figsize=(13,6))
dt = t2c.calc_dt(xf, dn, z)
ps, ks = t2c.power_spectrum_1d(dt, kbins=20, box_dims=box_len)
axs[0].pcolormesh(np.linspace(0,box_len,dt.shape[0]+1), 
				  np.linspace(0,box_len,dt.shape[1]+1), 
				  dt[:,:,0],
				  cmap='jet')
axs[0].set_xlabel('[Mpc]', fontsize=13)
axs[0].set_ylabel('[Mpc]', fontsize=13)
axs[1].loglog(ks, ps*ks**3/2/np.pi**2, lw=3, 
	label='$z={:.3f}$, $x_\mathrm{{HII}}={:.1f}$'.format(z,xf.mean()))
axs[1].legend(fontsize=14)
axs[1].set_xlabel('$k$ [1/Mpc]', fontsize=13)
axs[1].set_ylabel('$k^3 P/(2\pi^2)$ [mK$^2$]', fontsize=13)
plt.tight_layout()
plt.show()





src/simulations.py
import numpy as np 
from .functions import nearest_element_idx
from .download import * 
import tools21cm as t2c 

class C2RAY:
	'''
	The class to retrieve C2Ray simulations.
	
	'''
	def __init__(self, Nbody='CUBEP3M', work_dir='./', verbose=True):
		self.Nbody = Nbody
		self.verbose = verbose
		self.work_dir = work_dir 
		self.sim_file_list_dict = {}

	def set_nbody(self, nbody_url, info_type='dens', check_str='n_all.dat'):
		if self.Nbody.upper()=='CUBEP3M':
			nbody_file_list = get_file_full_list(nbody_url, ext='dat')
			file_list = []
			for ff in nbody_file_list: 
				if check_str in ff:
					file_list.append(ff)
			self.sim_file_list_dict[info_type] = file_list
		else:
			print('Method to retrieve {} outputs not implemented.'.format(nbody_file_list))

	def set_c2ray(self, c2ray_url, info_type='xfrac', check_str='xfrac3d_'):	# info_type='temp'
		c2ray_xfrac_file_list = get_file_full_list(c2ray_url, ext='bin')
		file_list = []
		for ff in c2ray_xfrac_file_list: 
			if check_str in ff:
				file_list.append(ff)
		self.sim_file_list_dict[info_type] = file_list

	def set_links(self, url_dict):
		zs_dict = {}
		if 'dens' in url_dict.keys(): 
			self.set_nbody(url_dict['dens'], info_type='dens', check_str='n_all.dat')
			zs_dict['dens'] = np.array([ff.split('/')[-1].split('n')[0] for ff in self.sim_file_list_dict['dens']]).astype(float)
		if 'xfrac' in url_dict.keys(): 
			self.set_c2ray(url_dict['xfrac'], info_type='xfrac', check_str='xfrac3d_')
			zs_dict['xfrac'] = np.array([ff.split('_')[-1].split('.b')[0] for ff in self.sim_file_list_dict['xfrac']]).astype(float)
		if 'temp' in url_dict.keys(): 
			self.set_c2ray(url_dict['temp'], info_type='temp', check_str='Temper3D_')
			zs_dict['temp'] = np.array([ff.split('_')[-1].split('.b')[0] for ff in self.sim_file_list_dict['temp']]).astype(float)
		if 'vel' in url_dict.keys():
			self.set_nbody(url_dict['vel'], info_type='vel', check_str='v_all.dat')
			zs_dict['vel'] = np.array([ff.split('/')[-1].split('v')[0] for ff in self.sim_file_list_dict['vel']]).astype(float)
		self.zs_dict = zs_dict

	def get_density_data_z(self, z):
		try:
			data_dir = self.work_dir['dens']
		except:
			data_dir = self.work_dir
		if not os.path.isdir(data_dir): os.mkdir(data_dir)
		if z in self.zs_dict['dens']:
			az = nearest_element_idx(self.zs_dict['dens'], z, both=False)
			zz = self.zs_dict['dens'][az]
			ff = download_simulation(self.sim_file_list_dict['dens'][az], data_dir, verbose=self.verbose)
			dn = t2c.DensityFile(ff).cgs_density
		else:
			az0, az1 = nearest_element_idx(self.zs_dict['dens'], z, both=True)
			zz0, zz1 = self.zs_dict['dens'][az0], self.zs_dict['dens'][az1]
			ff0 = download_simulation(self.sim_file_list_dict['dens'][az0], data_dir, verbose=self.verbose)
			ff1 = download_simulation(self.sim_file_list_dict['dens'][az1], data_dir, verbose=self.verbose)
			dn0 = t2c.DensityFile(ff0).cgs_density
			dn1 = t2c.DensityFile(ff1).cgs_density
			ff  = [ff0,ff1]
			dn  = dn0 + (dn1-dn0)*(z-zz0)/(zz1-zz0)
		return dn, ff

	def get_ionisation_data_z(self, z):
		try:
			data_dir = self.work_dir['xfrac']
		except:
			data_dir = self.work_dir
		if not os.path.isdir(data_dir): os.mkdir(data_dir)
		if z in self.zs_dict['xfrac']:
			az = nearest_element_idx(self.zs_dict['xfrac'], z, both=False)
			zz = self.zs_dict['xfrac'][az]
			ff = download_simulation(self.sim_file_list_dict['xfrac'][az], data_dir, verbose=self.verbose)
			xf = t2c.XfracFile(ff).xi 
		else:
			az0, az1 = nearest_element_idx(self.zs_dict['xfrac'], z, both=True)
			zz0, zz1 = self.zs_dict['xfrac'][az0], self.zs_dict['xfrac'][az1]
			ff0 = download_simulation(self.sim_file_list_dict['xfrac'][az0], data_dir, verbose=self.verbose)
			ff1 = download_simulation(self.sim_file_list_dict['xfrac'][az1], data_dir, verbose=self.verbose)
			xf0 = t2c.XfracFile(ff0).xi 
			xf1 = t2c.XfracFile(ff1).xi 
			ff  = [ff0,ff1]
			xf  = xf0 + (xf1-xf0)*(z-zz0)/(zz1-zz0)
		return xf, ff
		
	def get_data_z(self, z):
		data = {}

		xf, ff = self.get_ionisation_data_z(z)
		data['xfrac'] = xf 
		data['xfrac_filename'] = ff

		dn, ff = self.get_density_data_z(z)
		data['dens'] = dn 
		data['dens_filename'] = ff

		return data









examples/eor_hist.py
import numpy as np 
import matplotlib.pyplot as plt 
import pickle, os
from glob import glob
from StoReS import * 
import tools21cm as t2c 
from tqdm import tqdm 

box_len  = 244/0.7
save_dir = '../../'
eor_hist = {}

xf_url = 'https://ttt.astro.su.se/~gmell/244Mpc/LB1/'
dn_url = 'https://ttt.astro.su.se/~gmell/244Mpc/densities/nc250/coarser_densities/'
url_dict = {'xfrac': xf_url, 'dens': dn_url, }
c2r = C2RAY(work_dir=save_dir)
c2r.set_links(url_dict)
zs = np.intersect1d(c2r.zs_dict['xfrac'], c2r.zs_dict['dens'])
xs = []
for i,z in tqdm(enumerate(zs)):
	# z = 6.905
	data = c2r.get_data_z(z)
	xf = data['xfrac']
	dn = data['dens']
	xs.append(xf.mean())
	os.remove(data['xfrac_filename'])
	os.remove(data['dens_filename'])
xs = np.array(xs)
eor_hist['LB1'] = {'z': zs, 'x': xs}

fig, ax = plt.subplots(1,1,figsize=(6,5))
fig.suptitle('Dixon et al. (2016)')
ax.plot(eor_hist['LB1']['z'], eor_hist['LB1']['x'], lw=3, label='LB1')
ax.set_xlabel('z', fontsize=14)
ax.set_ylabel('$x_\mathrm{HII}$', fontsize=14)
ax.legend()
ax.axis([6,14,-0.01,1.01])
plt.tight_layout()
plt.show()


xf_url = 'https://ttt.astro.su.se/~gmell/244Mpc/LB2/'
dn_url = 'https://ttt.astro.su.se/~gmell/244Mpc/densities/nc250/coarser_densities/'
url_dict = {'xfrac': xf_url, 'dens': dn_url, }
c2r = C2RAY(work_dir=save_dir)
c2r.set_links(url_dict)
zs = np.intersect1d(c2r.zs_dict['xfrac'], c2r.zs_dict['dens'])
xs = []
for i,z in tqdm(enumerate(zs)):
	# z = 6.905
	data = c2r.get_data_z(z)
	xf = data['xfrac']
	dn = data['dens']
	xs.append(xf.mean())
	os.remove(data['xfrac_filename'])
	os.remove(data['dens_filename'])
xs = np.array(xs)
eor_hist['LB2'] = {'z': zs, 'x': xs}


xf_url = 'https://ttt.astro.su.se/~gmell/244Mpc/LB3/'
dn_url = 'https://ttt.astro.su.se/~gmell/244Mpc/densities/nc250/coarser_densities/'
url_dict = {'xfrac': xf_url, 'dens': dn_url, }
c2r = C2RAY(work_dir=save_dir)
c2r.set_links(url_dict)
zs = np.intersect1d(c2r.zs_dict['xfrac'], c2r.zs_dict['dens'])
xs = []
for i,z in tqdm(enumerate(zs)):
	# z = 6.905
	data = c2r.get_data_z(z)
	xf = data['xfrac']
	dn = data['dens']
	xs.append(xf.mean())
	os.remove(data['xfrac_filename'])
	os.remove(data['dens_filename'])
xs = np.array(xs)
eor_hist['LB3'] = {'z': zs, 'x': xs}


xf_url = 'https://ttt.astro.su.se/~gmell/244Mpc/LB4/'
dn_url = 'https://ttt.astro.su.se/~gmell/244Mpc/densities/nc250/coarser_densities/'
url_dict = {'xfrac': xf_url, 'dens': dn_url, }
c2r = C2RAY(work_dir=save_dir)
c2r.set_links(url_dict)
zs = np.intersect1d(c2r.zs_dict['xfrac'], c2r.zs_dict['dens'])
xs = []
for i,z in tqdm(enumerate(zs)):
	# z = 6.905
	data = c2r.get_data_z(z)
	xf = data['xfrac']
	dn = data['dens']
	xs.append(xf.mean())
	os.remove(data['xfrac_filename'])
	os.remove(data['dens_filename'])
xs = np.array(xs)
eor_hist['LB4'] = {'z': zs, 'x': xs}


fig, ax = plt.subplots(1,1,figsize=(7,6))
fig.suptitle('Dixon et al. (2016)')
ax.plot(eor_hist['LB1']['z'], 1-eor_hist['LB1']['x'], lw=3, ls='-',  label='LB1')
ax.plot(eor_hist['LB2']['z'], 1-eor_hist['LB2']['x'], lw=3, ls='--', label='LB2')
ax.plot(eor_hist['LB3']['z'], 1-eor_hist['LB3']['x'], lw=3, ls='-.', label='LB3')
ax.plot(eor_hist['LB4']['z'], 1-eor_hist['LB4']['x'], lw=3, ls=':',  label='LB4')
ax.set_xlabel('z', fontsize=14)
ax.set_ylabel('$x_\mathrm{HI}$', fontsize=14)
ax.legend()
ax.axis([6,14,-0.01,1.01])
plt.tight_layout()
plt.show()

